src/components/base/outline-modal/outline-modal.ts

```diff
@@ -0,0 +1,282 @@
+export type ModalSize = typeof modalSizes[number];
+
+// This is helpful in testing.
+export interface OutlineModalInterface extends HTMLElement {
+  isOpen: Boolean;
+  shouldForceAction: Boolean;
+  size?: ModalSize;
+  open: () => void;
+  close: () => void;
+}
+
+// See https://stackoverflow.com/questions/1599660/which-html-elements-can-receive-focus.
+// @todo make this re-usable across components?
+const focusableElementSelector = `
+  a[href]:not([tabindex="-1"]),
+  area[href]:not([tabindex="-1"]),
+  input:not([disabled]):not([tabindex="-1"]),
+  select:not([disabled]):not([tabindex="-1"]),
+  textarea:not([disabled]):not([tabindex="-1"]),
+  button:not([disabled]):not([tabindex="-1"]),
+  iframe:not([tabindex="-1"]),
+  [tabindex]:not([tabindex="-1"]),
+  [contentEditable=true]:not([tabindex="-1"])
+`;
+
+/**
+ * The Outline Modal component
+ * @element outline-modal
+ * @slot default - The modal contents
+ * @slot outline-modal--trigger - The trigger for the modal
+ * @slot outline-modal--header - The header in the modal
+ * @slot outline-modal--accessibility-description - The accessibility description which is used by screen readers.
+ */
+@customElement('outline-modal')
+export class OutlineModal
+  extends OutlineElement
+  implements OutlineModalInterface
+{
+  static styles: CSSResultGroup = [componentStyles];
+
+  @property({ attribute: false })
+  isOpen = false;
+
+  /**
+   * If we force the user to take an action, the consumer must provide a way to close the modal on their own.
+   */
+  @property({ type: Boolean })
+  shouldForceAction = false;
+
+  @property({ type: String })
+  size?: ModalSize = 'medium';
+
+  render(): TemplateResult {
+    return html`
+      <div
+        id="trigger"
+        tabindex="0"
+        @click="${this.open}"
+        @keydown="${this._handleTriggerKeydown}"
+      >
+        <slot name="outline-modal--trigger"></slot>
+      </div>
+      ${this._overlayTemplate()}
+    `;
+  }
+
+  @state()
+  _hasHeaderSlot: boolean;
+
+  @state()
+  _hasAccessibilityDescriptionSlot: boolean;
+
+  connectedCallback() {
+    super.connectedCallback();
+    this._handleSlotChange();
+  }
+
+  private _handleSlotChange(): void {
+    this._hasHeaderSlot =
+      this.querySelector('[slot="outline-modal--header"]') !== null;
+    this._hasAccessibilityDescriptionSlot =
+      this.querySelector(
+        '[slot="outline-modal--accessibility-description"]'
+      ) !== null;
+  }
+
+  private _overlayTemplate(): TemplateResult {
+    let template = html``;
+
+    if (this.isOpen) {
+      template = html`
+        <div
+          id="overlay"
+          tabindex="-1"
+          class="${this.size}"
+          @click="${this._handleOverlayClick}"
+          @keydown="${this._handleOverlayKeydown}"
+        >
+          <div
+            id="container"
+            role="dialog"
+            aria-modal="true"
+            aria-labelledby="${ifDefined(
```

Comment:
Let's move this conditional logic for `aria-labelledby` and `aria-described` into variables at the beginning of the render function to clean up this template a bit:
```
render() {
const ariaDesc = this._hasAccessibilityDescriptionSlot ? 'accessibility-description' : null
const ariaLabel = this._hasHeaderSlot ? 'header' : null
 ....
 <div aria-labeledby="${ifDefined(ariaLabel)} aria-describedby="${ifDefined(ariaDesc}">
....
}
```